base/random.jl

```diff
@@ -683,6 +683,44 @@ end
 rand{T}(rng::AbstractRNG, r::AbstractArray{T}, dims::Dims) = rand!(rng, Array{T}(dims), r)
 rand(rng::AbstractRNG, r::AbstractArray, dims::Int...) = rand(rng, r, dims)
 
+# rand from a string
+
+rand(rng::AbstractRNG, s::AbstractString) = rand_iter(rng, s, Base.iteratorsize(s))
+rand(s::AbstractString) = rand(GLOBAL_RNG, s)
+
+## optimization for String
+
+isvalid_unsafe(s::String, i) = !Base.is_valid_continuation(unsafe_load(pointer(s), i))
+
+function rand(rng::AbstractRNG, s::String)
+    rg = RangeGenerator(1:s.len)
+    while true
+        pos = rand(rng, rg)
+        isvalid_unsafe(s, pos) && return s[pos]
+    end
+end
+
+## rand from an iterator
+
+rand_iter(rng, s, ::Base.IteratorSize) = throw(ArgumentError("iterator must have a known length"))
+
+function rand_iter(rng::AbstractRNG, s, ::Union{Base.HasShape,Base.HasLength})::eltype(s)
```

Comment:
Yes Tony is spot on, and also initially I thought to propose this PR working generally on iterators, and the return type annotation could serve as an additional way to check whether the argument was complying to the iterator interface.